internal/venice-common/src/main/java/com/linkedin/venice/writer/VeniceWriter.java

```diff
@@ -1103,8 +1107,66 @@ private Future<RecordMetadata> putLargeValue(
       LeaderMetadataWrapper leaderMetadataWrapper,
       long logicalTs,
       Optional<PutMetadata> putMetadata) {
-    int replicationMetadataPayloadSize = putMetadata.isPresent() ? putMetadata.get().getSerializedSize() : 0;
-    int sizeAvailablePerMessage = maxSizeForUserPayloadPerMessageInBytes - serializedKey.length;
+    int replicationMetadataPayloadSize = putMetadata.map(PutMetadata::getSerializedSize).orElse(0);
+    final Supplier<String> reportSizeGenerator =
+        () -> getSizeReport(serializedKey.length, serializedValue.length, replicationMetadataPayloadSize);
+    ChunkedPayloadAndManifest valueChunksAndManifest =
+        chunksPayloadAndSend(serializedKey, serializedValue, valueSchemaId, callback, partition, reportSizeGenerator);
+    ChunkedPayloadAndManifest rmdChunksAndManifest = null;
+    if (isRmdChunkingEnabled) {
+      rmdChunksAndManifest =
+          chunksPayloadAndSend(serializedKey, serializedValue, valueSchemaId, callback, partition, reportSizeGenerator);
+    }
+    // Now that we've sent all the chunks, we can take care of the final value, the manifest.
```

Comment:
Are we able to abstract some changes in `ChunkAssembler` such that we can directly use it here?

Reply:
After checking this class, this is the assembling chunks instead of splitting chunks, so it should not be there. I will try to see if I can try to abstract it somewhere.

Reply:
I made an attempt in the latest commit to move this part of logic into a standalone helper class and add a unit test for it, please feel free to double check.